### PR TITLE
fix protocol generate problem when using cpp

### PR DESCRIPTION
**Pull request by luohaha**
Because I want to build cpp application based on tendermint abci，so I use these protocol files to generate grpc and pb cpp files. But I meet these:
```
../../tendermint/consensus/types.pb.h:3659:57: error: return type of out-of-line definition of
      'tendermint::consensus::Message::has_vote' differs from that in the declaration
inline const ::tendermint::consensus::HasVote& Message::has_vote() const {
             ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~          ^
../../tendermint/consensus/types.pb.h:3553:22: note: previous definition is here
inline bool Message::has_vote() const {
```

`vote` and `has_vote` will cause function name conflict in pb.h file. So I solve this problem with this easy way.

### CODE REVIEW

File: proto/tendermint/consensus/types.proto [hunk 1]
```diff
@@ -85,7 +85,7 @@ message Message {
     ProposalPOL   proposal_pol    = 4;
     BlockPart     block_part      = 5;
     Vote          vote            = 6;
-    HasVote       has_vote        = 7;
+    HasVote       have_vote       = 7;
```
Comment (marbar3778) [inline proto/tendermint/consensus/types.proto:88]: We will need to regenerate the protobuf files to accommodate this change. Is there any way cpp could disable generating the code that causes the issue?

Comment (luohaha) [reply proto/tendermint/consensus/types.proto:88]: I afraid can't do that, because that is Protobuf rules. How about I rename some variable in message.go?

**Comment by marbar3778** [comment]
Hey, it may be better to rename all the messages. Having one be `has` and another be `have` is a bit messy. Also you will need to regenerate and change things in the go code as well.

**Comment by luohaha** [comment]
> Hey, it may be better to rename all the messages. Having one be `has` and another be `have` is a bit messy. Also you will need to regenerate and change things in the go code as well.

Done.

**Comment by luohaha** [comment]
Sorry for delay so long

**Comment by marbar3778** [comment]
One last nit, could you modify the spec as well? (https://github.com/tendermint/spec/tree/master/proto/consensus) This will keep the spec compatible with the implementation.

**Comment by luohaha** [comment]
> One last nit, could you modify the spec as well? (https://github.com/tendermint/spec/tree/master/proto/consensus) This will keep the spec compatible with the implementation.

https://github.com/tendermint/spec/pull/289

**Review by tessr — CHANGES_REQUESTED** [review]
Hi @luohaha, 

Apologies for jumping in here so late, but  `HaveVote` is grammatically confusing. How about `ReceivedVote` instead? (Thanks @cmwaters for the idea!)

**Comment by luohaha** [comment]
> Hi @luohaha,
> 
> Apologies for jumping in here so late, but `HaveVote` is grammatically confusing. How about `ReceivedVote` instead? (Thanks @cmwaters for the idea!)

ReceivedVote is better. Done.

**Comment by cmwaters** [comment]
Hey @luohaha, thanks for making the changes. A few of us are away at the moment so we will look to address this and hopefully get it merged next week. Just wanted to give you a heads up.

**Comment by tessr** [comment]
This also needs a changelog entry, but I've dismissed my "needs changes" review and I believe it can be merged once the changelog has been updated.

**Comment by marbar3778** [comment]
@luohaha what is the reason you are generating the consensus message proto files? If you are building an app with abci this is not needed.

**Comment by marbar3778** [comment]
@luohaha I'll leave this open for a few more days, but after that I will close. For building apps, please only compile the ABCI proto files.

**Comment by luohaha** [comment]
> @luohaha what is the reason you are generating the consensus message proto files? If you are building an app with abci this is not needed.

I meet this problem when I do some hacking with tendermint core. And protobuf file should be compatible in multi-language environment.

**Comment by luohaha** [comment]
@marbar3778 When you want to rewrite tendermint node in cpp，and communicate with go node/rust node, that will be a problem.

**Comment by marbar3778** [comment]
Looking into protoc cpp to read if they mark words as do not use in the naming. 

> @marbar3778 When you want to rewrite tendermint node in cpp，and communicate with go node/rust node, that will be a problem.

we don't have any plans for a cpp node, will double check what it says for cpp generator. Otherwise we will need to make changes in the spec to have them land in the implementation

**Comment by marbar3778** [comment]
@luohaha are you in our discord? we can chat there to avoid going back and forth in github comments.

**Comment by marbar3778** [comment]
```
The compiler will generate the following accessor methods:

    bool has_foo() const: Returns true if the field is set. 
```

Ill quickly add a testvector on master to make sure the encoding stays the same. If you are building an app then this isn't urgent, right?

**Comment by luohaha** [comment]
> @luohaha are you in our discord? we can chat there to avoid going back and forth in github comments.

Not in discord yet, please give me the address.

**Comment by marbar3778** [comment]
> Not in discord yet, please give me the address.

https://discord.gg/vcExX9T

**Comment by marbar3778** [comment]
Hey, I added some test vectors, could you update them accordingly. Once the test passes we can move forward with the merge here

**Pull request closed without merging**